client/blocks/qr-code-login/index.jsx

```diff
@@ -0,0 +1,266 @@
+		anon_id: anonymousUserId,
+		token,
+		data: encrypted,
+	} );
+
+	if ( responseData === false ) {
+		setIsErrorState( true );
+		setPullInterval( null );
+		return;
+	}
+
+	setAuthState( responseData.data );
+};
+
+function TokenQRCode( { tokenData } ) {
+	if ( ! tokenData ) {
+		return <QRCodePlaceholder />;
+	}
+	const { token, encrypted } = tokenData;
+	const imageSettings = {
+		src: qrCenter,
+		height: 64,
+		width: 64,
+		excavate: false,
+	};
+	return (
+		<QRCode
+			value={ `https://apps.wordpress.com/get/?campaign=login-qr-code#qr-code-login?token=${ token }&data=${ encrypted }` }
+			size={ 300 }
+			imageSettings={ imageSettings }
+		/>
+	);
+}
+
+function QRCodePlaceholder() {
+	return (
+		<div className="qr-code-login__placeholder">
+			<span className="qr-code-login__corner-box"></span>
+			<span className="qr-code-login__corner-box"></span>
+			<span className="qr-code-login__corner-box"></span>
+		</div>
+	);
+}
+
+function QRCodeLogin() {
+	const translate = useTranslate();
+	const [ tokenData, setTokenData ] = useState( null );
+	const [ authState, setAuthState ] = useState( false );
+	const [ pullInterval, setPullInterval ] = useState( AUTH_PULL_INTERVAL );
+	const [ isErrorState, setIsErrorState ] = useState( false );
+
+	const anonymousUserId = getTracksAnonymousUserId();
+
+	// Set the error state if we don't have a anonymousUserId
+	useEffect( () => {
+		if ( ! anonymousUserId ) {
+			setIsErrorState( true );
+		}
+	}, [ anonymousUserId, setIsErrorState ] );
+
+	// Fetch QR code data.
+	useEffect( () => {
+		fetchQRCodeData( setTokenData, tokenData, anonymousUserId, setIsErrorState );
+	}, [ setTokenData, tokenData, anonymousUserId, setIsErrorState ] );
+
+	// Fetch the Auth Data.
+	useInterval( () => {
+		fetchAuthState(
+			setAuthState,
+			setTokenData,
+			tokenData,
+			anonymousUserId,
+			isErrorState,
+			setIsErrorState,
+			setPullInterval
+		);
+	}, pullInterval );
+
+	// Send the user to the login state.
+	useEffect( () => {
+		if ( authState?.auth_url ) {
+			window.location.replace( authState.auth_url );
+		}
+	}, [ authState ] );
+
+	useEffect( () => {
+		getStoredItem( LOCALE_STORAGE_KEY ).then( ( localTokenData ) =>
+			setTokenData( localTokenData ?? false )
+		);
+	}, [] );
+
+	const steps = useMemo( () => [
+		// translation: Link to the WordPress App.
+		translate( 'Open the {{link}}%(wordpress)s App{{/link}} on your phone.', {
+			args: {
+				wordpress: 'WordPress',
+			},
+			components: {
+				link: (
+					<ExternalLink
+						target="_blank"
+						icon={ false }
+						href="https://apps.wordpress.com/get/?campaign=calypso-qrcode-apps"
+					/>
+				),
+			},
+		} ),
+		translate( 'Tap the My Site Tab.' ),
+		translate( 'Tap your Profile image in the top right corner of the screen.' ),
+		translate( 'Tap the {{strong}}Scan QR Code{{/strong}} option.', {
+			components: {
+				strong: <strong />,
+			},
+		} ),
+		translate( 'Point your phone to this screen to scan the code.' ),
+	] );
```

Comment:
Should we add `translate` as a dependency? The `translate` function will be redefined if the locale changes, which is when we'd want to recompute these localized values.

Reply:
I don't quite understand this suggestion. Do you have an example of this in a different component? 
What benefit do you see as a result?

Reply:
Calling `useMemo()` without any dependencies (the second argument of `useMemo`) will memoize the value. That's fine and expected, of course. 

However, if the locale changes, those strings may not change to be in the current language, and I'd expect that they change. To fix that, we can add `[ translate ]` as the second argument of `useMemo()`. That way, because `translate()` gets rebound on a locale change, the component will be re-rendered and the `useMemo` value will be recomputed to call `translate()` with the new locale. 

As for examples, just look at some `useMemo()` usages and you'll see that they often provide dependencies.

Just noting that without any dependencies (like it is now), `useMemo()` will be recomputed on every re-render, which makes it a bit useless right now. If you pass an empty array for dependencies, it would be computed just once and never recomputed again. If you pass `[ translate ]`, it will recompute if the locale changes, which is the expected behavior.

Does that make sense?

Reply:
I personally wouldn't call `useMemo` here at all. It doesn't provide any meaningful optimization.

Reply:
I removed the useMemo call here.

Reply:
@tyxla thanks for your explanation. 

> Does that make sense?
It does make more sense now. I didn't know that this will not work for users that switch languages.